xcm/xcm-simulator/example/src/lib.rs

```diff
@@ -182,33 +192,201 @@ mod tests {
+				max_weight_for_execution,
 			));
+			let para_account_a = ParaId::from(1).into_account();
+			assert_eq!(
+				parachain::Balances::free_balance(&para_account_a),
+				INITIAL_BALANCE + withdraw_amount
+			);
 		});
 
-		ParaA::execute_with(|| {
-			// free execution, full amount received
+		ParaA::execute_and_dispatch_xcm(|| {
+			// Check message received
+			let expected_message = (
+				X1(Parent),
+				ReserveAssetDeposit {
+					assets: vec![ConcreteFungible { id: X1(Parent), amount: withdraw_amount }],
+					effects: vec![
+						buy_execution(max_weight_for_execution),
+						Order::DepositAsset {
+							assets: vec![All],
+							dest: X1(Junction::AccountId32 {
+								network: NetworkId::Any,
+								id: ALICE.into(),
+							}),
+						},
+					],
+				},
+			);
+			assert_eq!(parachain::MsgQueue::received_dmp(), vec![expected_message]);
+			// Check message execution with full amount received
 			assert_eq!(
 				pallet_balances::Pallet::<parachain::Runtime>::free_balance(&ALICE),
-				INITIAL_BALANCE + 123
+				INITIAL_BALANCE + withdraw_amount
+			);
+		});
+	}
+
+	/// Scenario:
+	/// A parachain transfers funds on the relay chain to another parachain account.
+	///
+	/// Asserts that the parachain accounts are updated as expected.
+	#[test]
+	fn withdraw_and_deposit() {
+		MockNet::reset();
+
+		let send_amount = 10;
+		let mut amount_received = 0;
+		let weight_for_execution = 3 * relay_chain::BaseXcmWeight::get();
+
+		ParaA::execute_and_dispatch_xcm(|| {
+			let message = WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount: send_amount }],
+				effects: vec![
+					buy_execution(weight_for_execution),
+					Order::DepositAsset { assets: vec![All], dest: Parachain(2).into() },
+				],
+			};
+			// Send withdraw and deposit
+			assert_ok!(ParachainPalletXcm::send_xcm(Null, X1(Parent), message.clone(),));
+		});
+
+		amount_received += send_amount;
+
+		Relay::execute_and_dispatch_xcm(|| {
+			let para_account_a: relay_chain::AccountId = ParaId::from(1).into_account();
+			assert_eq!(
+				relay_chain::Balances::free_balance(para_account_a),
+				INITIAL_BALANCE - send_amount
+			);
+			let para_account_b: relay_chain::AccountId = ParaId::from(2).into_account();
+			assert_eq!(relay_chain::Balances::free_balance(para_account_b), amount_received);
+		});
+	}
+
+	/// Scenario:
+	/// A parachain wants to be notified that a transfer worked correctly.
+	/// It sends a `QueryHolding` after the deposit to get notified on success.
+	///
+	/// Asserts that the balances are updated correctly and the expected XCM is sent.
+	#[test]
+	fn query_holding() {
+		MockNet::reset();
+
+		let send_amount = 10;
+		let mut amount_spent = 0;
+		let mut amount_received = 0;
+		let weight_for_execution = 3 * relay_chain::BaseXcmWeight::get();
+		let query_id_set = 1234;
+
+		// First send a message which fails on the relay chain
+		ParaA::execute_and_dispatch_xcm(|| {
```

Comment:
this failure is not part of the test dsecription.

Reply:
removed it, I agree